scheduler/src/cook/mesos/data_locality.clj

```diff
@@ -1,46 +1,148 @@
+            [cook.config :as config]
+            [cook.mesos.util :as util]
+            [plumbing.core :as pc]
+            [clojure.tools.logging :as log]
+            [datomic.api :as d])
   (:import com.netflix.fenzo.VMTaskFitnessCalculator
-           com.netflix.fenzo.plugins.BinPackingFitnessCalculators))
+           com.netflix.fenzo.plugins.BinPackingFitnessCalculators
+           java.util.UUID))
+
+(let [datasets->host-name->cost-atom (atom {})
+      datasets->last-update-time-atom (atom {})]
+
+  (defn reset-data-local-costs! []
+    (reset! datasets->host-name->cost-atom {})
+    (reset! datasets->last-update-time-atom {}))
 
-(let [data-local-costs (atom {})]
   (defn update-data-local-costs
     "Updates the current data local costs. Costs larger than the configured max-cost will be capped at max-cost
      and negative costs will be reset to 0."
-    [job->host->cost]
-    (let [{:keys [maximum-cost]} (config/data-local-fitness-config)]
-      (->> job->host->cost
-           (pc/map-vals (fn [host->cost]
-                          (pc/map-vals (fn [cost] (-> cost (min maximum-cost) (max 0)))
-                                       host->cost)))
-           (reset! data-local-costs))))
+    [datasets->host->cost datasets-to-remove]
+    (let [{:keys [maximum-cost]} (config/data-local-fitness-config)
+          clean-datasets->host->cost (pc/map-vals (fn [host->cost]
+                                                   (pc/map-vals (fn [cost] (-> cost (min maximum-cost) (max 0)))
+                                                                host->cost))
+                                                 datasets->host->cost)]
+      (swap! datasets->host-name->cost-atom (fn update-datasets->host-name->cost-atom-atom
+                                [current]
+                                (let [remove-old-values (apply dissoc current datasets-to-remove)]
+                                  (merge remove-old-values clean-datasets->host->cost))))
+      (swap! datasets->last-update-time-atom (fn update-last-update-time
+                                [current]
+                                (let [remove-old-values (apply dissoc current datasets-to-remove)
+                                      new-values (pc/map-from-keys (constantly (t/now)) (keys clean-datasets->host->cost))]
+                                  (merge remove-old-values new-values))))))
 
   (defn get-data-local-costs
     "Returns the current cost for jobs to run on each host"
     []
-    @data-local-costs))
+    @datasets->host-name->cost-atom)
+
+
+  (defn get-last-update-time
+    "Returns the last update time information"
+    []
+    @datasets->last-update-time-atom)
+
+  (defn jobs-to-update
+    "Return the list of job ids to update data locality costs for in the next iteration.
+     The number of jobs is limited by the configured batch size."
+    [db]
+    (let [{:keys [batch-size]} (config/data-local-fitness-config)
```

Comment:
There's a danger here, depending on how many jobs we have in the queue, how long a fetch cycle takes, and the batch size, we could be running off data of a variable staleness. We also most want *fresh* locality information from jobs that are near the top of the queue. I'm not proposing solving this problem now, but IMO, we should measure it.

At an appropriate place in the scheduler, where we use locality information to schedule (Two points, when we launch jobs, and when they're in the part of the DRU queue we ship to fenzo), we should log a two metrics 'locality-staleness-launch', and 'locality-stalenss-considerable'. Track the metric, report it to our business partners and see if the numbers are acceptable.

Reply:
I'm a little hesitant to add a bunch of metric updates to the match process, instead I added one to update the staleness of all of the cached data as part of the update cycle. How does that work for you?